components/menu/menu-item-mixin.js

```diff
@@ -0,0 +1,146 @@
+export const MenuItemMixin = superclass => class extends superclass {
+
+	static get properties() {
+		return {
+			disabled: { type: Boolean, reflect: true },
+			first: { type: Boolean, reflect: true }, // set by d2l-menu
+			hasChildView: { type: Boolean, attribute: 'has-child-view' },
+			hidden: { type: Boolean, reflect: true },
+			last: { type: String, reflect: true }, // set by d2l-menu
+			role: { type: String, reflect: true },
+			tabindex: { type: Number, reflect: true, attribute: 'tab-index' },
+			text: String
+		};
+	}
+
+	constructor() {
+		super();
+		this.__keyCodes = {
+			ENTER: 13,
+			LEFT: 37,
+			RIGHT: 39,
+			SPACE: 32
+		};
+		this.__children = null;
+		this.role = 'menuitem';
+		this.tabindex = -1;
+	}
+
+	firstUpdated() {
+		super.firstUpdated();
+		this.__initializeItem();
+	}
+
+	updated(changedProperties) {
+		super.updated(changedProperties);
+
+		changedProperties.forEach((oldValue, propName) => {
+			if (propName === 'hidden') {
+				this._onHidden();
+			}
+		});
+	}
+
+	connectedCallback() {
+		super.connectedCallback();
+		this.addEventListener('blur', this.__onBlur);
+		this.addEventListener('click', this.__onClick);
+		this.addEventListener('d2l-hierarchical-view-hide-complete', this.__onHideComplete);
+		this.addEventListener('dom-change', this.__onDomChange);
+		this.addEventListener('focus', this.__onFocus);
+		this.addEventListener('keydown', this.__onKeyDown);
+	}
+
+	disconnectedCallback() {
+		super.disconnectedCallback();
+		this.removeEventListener('blur', this.__onBlur);
+		this.removeEventListener('click', this.__onClick);
+		this.removeEventListener('d2l-hierarchical-view-hide-complete', this.__onHideComplete);
+		this.removeEventListener('dom-change', this.__onDomChange);
+		this.removeEventListener('focus', this.__onFocus);
+		this.removeEventListener('keydown', this.__onKeyDown);
+	}
+
+	__initializeItem() {
+		const children = Array.from(this.children);
+		if (children && children.length > 0 && children[0].tagName === 'TEMPLATE') {
+			return;
+		}
+
+		for (let i = 0; i < children.length; i++) {
+			if (children[i].tagName !== 'TEMPLATE' && children[i].tagName !== 'DOM-IF') {
```

Comment:
Wonder what we should do here about `dom-if` since that's a Polymer thing. I guess technically this could still get used with Polymer, but it's a bit strange for it to "know" about it.

Reply:
Yeah that's not amazing eh... I'm not sure if we can safely ignore that case.